### PR TITLE
Change payment method

### PR DESCRIPTION
**Pull request by valentinludu**
### <samp>🤖 Generated by Copilot at 03e7a8d</samp>

This pull request adds the ability for users to change their card details for space subscriptions within the app, using Stripe elements and the `SubscriptionApi` class. It also simplifies and refactors some of the existing code related to subscriptions and payment methods, and removes the unused billing portal feature. It updates the `stripe` packages and the `stripeMock` module, and adds new files and tests for the new functionality.

### WHY
Changes:
- removed billing portal implementation
- added stripe card elements for changing the credit card details of a subscription
- only after confirmation of the card we attach it to the customer and subscription
- make a general confirmation modal for multiple use cases
- fix email not updating
- improved email UX -> email can't be updated if the field is the same with the one from stripe

New flow added in Miro
https://miro.com/app/board/uXjVME0MLtU=/

### CODE REVIEW

File: components/common/Modal/ConfirmModal.tsx [hunk 1]
```diff
@@ -10,25 +10,25 @@ type Props = Pick<ModalProps, 'onClose' | 'open' | 'size'> & {
...
   title?: string;
-  onConfirm: () => void;
+  onConfirm: () => Promise<void> | void;
   secondaryButtonText?: string;
   onClose: () => void;
   disabled?: boolean;
 };
 
-export default function ConfirmUpgradeModal({
+export default function ConfirmModal({
```
Comment (mattcasey) [inline components/common/Modal/ConfirmModal.tsx:19]: thx for the cleanup

File: pages/api/spaces/[id]/payment-method.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    throw new NotFoundError(`Subscription not found for space ${spaceId}`);
+  }
+
+  await updatePaymentMethod({
+    customerId: subscriptionData.customerId,
+    subscriptionId: subscriptionData.subscriptionId,
+    paymentMethodId
+  });
+
+  log.info(`Payment method has been updated for space ${spaceId} by the user ${userId}`);
```
Comment (mattcasey) [inline pages/api/spaces/[id]/payment-method.ts:59]: spaceId and userId should be meta properties of the log

File: components/settings/subscription/ChangeCardDetails.tsx [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    if (!stripe || !elements) {
+      return;
+    }
+
+    const cardNumber = elements.getElement(CardNumberElement);
+    const cardExpiry = elements.getElement(CardExpiryElement);
+    const cardCvc = elements.getElement(CardCvcElement);
+    const cardError = !cardNumber || !cardExpiry || !cardCvc;
+
+    if (cardError) {
```
Comment (mattcasey) [inline components/settings/subscription/ChangeCardDetails.tsx:68]: not really an error, i might suggest instead a variable name like "isMissingDetails"

File: components/settings/subscription/ChangeCardDetails.tsx [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+  const handleCardDetails = (disabled: boolean) => {
+    setIsDisabled(disabled);
+  };
+
+  return (
+    <>
+      <Button {...bindTrigger(updatePaymentMethodPopup)} variant='text' sx={{ px: 0 }}>
+        Update your payment details
+      </Button>
+      <ConfirmModal
```
Comment (mattcasey) [inline components/settings/subscription/ChangeCardDetails.tsx:121]: For the future: the 'ConfirmModal' is so non-specific it just seems like an alternative for Modal. Changing settings is not a 'confirmation' for example. A confirmation should be yes/no question. Maybe its "ModalWithButtons" and "question" is just called "content" or even replaced by "children".

**Review by mattcasey — APPROVED** [review]

File: pages/api/spaces/[id]/payment-method.ts (hunk 2 above)
Comment (valentinludu) [reply pages/api/spaces/[id]/payment-method.ts:59]: Done

**Pull request merged**